src/frontend/src/components/authenticateBox.ts

```diff
@@ -45,52 +45,48 @@ type ChasmOpts = {
 
 /** Authentication box component which authenticates a user
  * to II or to another dapp */
-export const authenticateBox = (
+export const authenticateBox = async (
   connection: Connection,
   templates: AuthTemplates
 ): Promise<LoginData> => {
-  const retryOnError = async (result: LoginFlowResult): Promise<LoginData> => {
+  const promptAuth = () =>
+    new Promise<LoginFlowResult>((resolve) => {
+      authenticateBoxTemplate({
+        templates,
+        addDevice: () => addRemoteDevice(connection),
+        onContinue: (userNumber) => {
+          resolve(authenticate(connection, userNumber));
+        },
+        register: () => {
+          resolve(registerIfAllowed(connection));
+        },
+        recoverAnchor: (userNumber) => useRecovery(connection, userNumber),
+        userNumber: getUserNumber(),
+      });
+    });
+
+  // Retry until user has successfully authenticated
+  for (;;) {
+    const result = await promptAuth();
     switch (result.tag) {
+      case "ok":
+        setUserNumber(result.userNumber);
+        return result;
       case "err":
         await displayError({
           title: result.title,
           message: result.message,
           detail: result.detail !== "" ? result.detail : undefined,
           primaryButton: "Try again",
         });
-        return authenticateBox(connection, templates);
-      case "ok":
-        return result;
+        break;
       case "canceled":
-        return authenticateBox(connection, templates);
+        break;
       default:
         unreachable(result);
         break;
```

Comment:
shouldn't  this `break` becomes now a `return` otherwise it might continue looping if `default` is reached which was not the case before?

Reply:
`default` is actually never reached, that's what `unreachable` ensures. A `return` here would actually break the flow and make the app non-responsive because the promise is never resolved